renderdoc/driver/vulkan/vk_initstate.cpp

```diff
@@ -1975,28 +1975,78 @@ void WrappedVulkan::Apply_InitialState(WrappedVkRes *live, VkInitialContents ini
+    if(!memRefs)
+    {
+      // No information about the memory usage in the frame.
+      // Pessimistically assume the entire memory needs to be reset.
+      resetReq.update(0, initial.mem.size, eInitReq_Reset,
+                      [](InitReqType x, InitReqType y) -> InitReqType { return std::max(x, y); });
+    }
+    else
+    {
+      bool initialized = memRefs->initializedLiveRes == live;
+      memRefs->initializedLiveRes = live;
+      for(auto it = memRefs->rangeRefs.begin(); it != memRefs->rangeRefs.end(); it++)
+      {
+        InitReqType t = InitReq(it->value());
+        if(t == eInitReq_Reset || (t == eInitReq_InitOnce && !initialized))
+          resetReq.update(it->start(), it->finish(), eInitReq_Reset,
+                          [](InitReqType x, InitReqType y) -> InitReqType { return std::max(x, y); });
+        else if(t == eInitReq_Clear || (t == eInitReq_None && !initialized))
+          resetReq.update(it->start(), it->finish(), eInitReq_Clear,
+                          [](InitReqType x, InitReqType y) -> InitReqType { return std::max(x, y); });
+      }
+    }
+
     VkResult vkr = VK_SUCCESS;
 
     VkCommandBufferBeginInfo beginInfo = {VK_STRUCTURE_TYPE_COMMAND_BUFFER_BEGIN_INFO, NULL,
                                           VK_COMMAND_BUFFER_USAGE_ONE_TIME_SUBMIT_BIT};
 
     VkBuffer srcBuf = initial.buf;
-    VkDeviceSize datasize = initial.mem.size;
-    VkDeviceSize dstMemOffs = 0;
+
+    VkBuffer dstBuf = m_CreationInfo.m_Memory[id].wholeMemBuf;
+    if(dstBuf == VK_NULL_HANDLE)
+    {
+      RDCERR("Whole memory buffer not present for %llu", id);
+      return;
+    }
+
+    if(resetReq.size() == 1 && resetReq.begin()->value() == eInitReq_None)
+      return;    // no copy or clear required
 
     VkCommandBuffer cmd = GetNextCmd();
 
     vkr = ObjDisp(cmd)->BeginCommandBuffer(Unwrap(cmd), &beginInfo);
     RDCASSERTEQUAL(vkr, VK_SUCCESS);
 
-    VkBuffer dstBuf = m_CreationInfo.m_Memory[id].wholeMemBuf;
-
-    VkBufferCopy region = {0, dstMemOffs, datasize};
-
-    if(dstBuf != VK_NULL_HANDLE)
-      ObjDisp(cmd)->CmdCopyBuffer(Unwrap(cmd), Unwrap(srcBuf), Unwrap(dstBuf), 1, &region);
-    else
-      RDCERR("Whole memory buffer not present for %llu", id);
+    std::vector<VkBufferCopy> regions;
+    uint32_t fillCount = 0;
+    for(auto it = resetReq.begin(); it != resetReq.end(); it++)
+    {
+      if(it->start() >= initial.mem.size)
+        continue;
+      VkDeviceSize finish = RDCMIN(it->finish(), initial.mem.size);
+      VkDeviceSize size = finish - it->start();
+      switch(it->value())
+      {
+        case eInitReq_Clear:
+          ObjDisp(cmd)->CmdFillBuffer(Unwrap(cmd), Unwrap(dstBuf), it->start(), size, 0);
+          fillCount++;
+          break;
+        case eInitReq_Reset: regions.push_back({it->start(), it->start(), size}); break;
+        default: break;
+      }
+    }
+    RDCLOG("Apply_InitialState: %d fills, %d copies", fillCount, regions.size());
```

Comment:
I could see this log message getting a bit spammy since this will be hit every replay operation for every memory object - I'd suggest demoting it to `RDCDEBUG` so it doesn't show up in release builds as it's likely only useful when debugging renderdoc with a repro case.

It might also be useful to print the original ID for the memory object so we can tie together these stats with the objects in the capture.

Reply:
Yes, I put that in while I was testing, and did not intend to leave it in at all. But I agree that this could be useful for debugging, particularly to highlight cases that cause many small copies or clears, so I will leave it in as `RDCDEBUG`.